pkg/roll/roll.go

```diff
@@ -162,3 +168,56 @@ func (m *Roll) Close() error {
 
 	return m.pgConn.Close()
 }
+
+// UnappliedMigrations returns a slice of unapplied migrations from `dir`,
+// lexicographically ordered by filename. Applying each of the returned
+// migrations in order will bring the database up to date with `dir`.
+func (m *Roll) UnappliedMigrations(ctx context.Context, dir fs.FS) ([]*migrations.Migration, error) {
+	latestVersion, err := m.State().LatestVersion(ctx, m.Schema())
+	if err != nil {
+		return nil, fmt.Errorf("determining latest version: %w", err)
+	}
+
+	files, err := fs.Glob(dir, "*.json")
+	if err != nil {
+		return nil, fmt.Errorf("reading directory: %w", err)
+	}
+
+	// Find the index of the first unapplied migration
+	var idx int
+	if latestVersion != nil {
+		for _, file := range files {
+			file, err := dir.Open(file)
+			if err != nil {
+				return nil, fmt.Errorf("opening migration file %q: %w", file, err)
+			}
+
+			migration, err := migrations.ReadMigration(file)
+			if err != nil {
+				return nil, fmt.Errorf("reading migration file %q: %w", file, err)
+			}
+
+			idx++
+			if migration.Name == *latestVersion {
+				break
+			}
```

Comment:
This could occur if a user runs `pgroll migrate foo/` where `foo` is a directory containing migrations other than the set applied to the target database.

This is a specific case of the following behaviour:

If I have a target schema at version `03_foo` and I have a `migrations/` directory like this:

```
01_apples.json
02_bananas.json
03_foo.json
04_cherries.json
```

`pgroll migrate migrations/` will apply `04_cherries.json` even if `01_apples` and `02_bananas` were never applied to the target schema; all migrations from the latest version of the target schema onwards are applied without checking if versions prior to `03_foo` were applied.

For the common case where a user is applying the right migrations directory this behaviour is fine, but we could be more strict and fail if the exact sequence of migrations on disk has not been applied.

WDYT?

Reply:
yes, I believe that would be my expectation as a user?

Reply:
This is implemented and tested in 3c5824caa7bb3ac9972e82eddd4172409b5dea78.

If the local migration directory doesn't match the remote history, `Roll.UnappliedMigrations` now returns an `ErrMismatchedMigration`.